### ISSUE TITLE
provide automatic "sensible" number of subsets for OS algorithms

### ISSUE THREAD

**Issue by KrisThielemans** [UCL/STIR#741]
It would be convenient to be able to use `.par` files that work for all (or at least most) scanners, e.g. [for scatter](https://github.com/UCL/STIR/pull/740#issue-517716191). For OS algorithms like `OSEM`, `OSMAPOSL` and even `OSSPS`, the admissible number of subsets is determined by the number of views and symmetry settings. Possibly we could provide a new keyword for these algorithms (or is it in `GeneralisedObjectiveFunction`?) like
```
choose number of subsets automatically := 1
```
or (a bit harder for the parsing, but doable)
```
number of subsets:=automatic
```
We would then need to have some rule to determine this, but I'd suggest to keep this really simple, e.g. not depend on counts etc, but just on geometry.

@robbietuk what do you think?

**Comment by robbietuk** [comment]
It is a great idea, but how would you suggest handling `number of subiterations`? These two variables are fairly tied together and it is very difficult to suggest an automatic `number of subiterations`.

Instead we could provide example parameter files for a few "real" scanners with a reasonable number of subsets and subiterations.

**Comment by KrisThielemans** [comment]
I don't agree. People tend to take the `number of subiterations`  as a multiple of the number of subsets but there's no reason for that. What matters for an application is the number of updates that happen.

**Comment by robbietuk** [comment]
This is true, generally people actually want (5) epochs of processing. 24 subsets therefore needs neeed 24 * 5 subiterations. 

I think what you are suggesting is doable. Regarding rules, I would suggest we aim for 10 subsets by defaultand scale the number dependant on the factors of the number of projection angles (when considering symmetries). 

Examples: D690 (288 views) with 4 symmetries -> 72 symmetric projections -> 9/12 subsets.

**Comment by KrisThielemans** [comment]
> generally people actually want (5) epochs of processing. 24 subsets therefore needs need 24 * 5 subiterations.

I don't think so, they want 120 updates 😄 

not sure what you mean with 9/12, but certainly doable

**Comment by robbietuk** [comment]
I personally think that computing time is the limiting factor. Doing a reconstruction using 5 subsets with 120 updates is a lot of processing (24 epochs). As you stated "_...the number of subiterations as a multiple of the number of subsets..._", indicating that people want `n` number of epochs of processing. 
This is why I perfer to focus on the number of epochs in algorithm assessment when comparing different numbers of subsets reconstruction.

>not sure what you mean with 9/12[...]

Sorry, I meant 9 or 12 subsets. Keep the number of subsets as close to 10 as possible.

**Comment by KrisThielemans** [comment]
> This is why I perfer to focus on the number of epochs in algorithm assessment when comparing different numbers of subsets reconstruction.

sure, but when providing default options, we're wanting to make life easy for people. w/hen using OSEM in real life, you shoudl think about how many updates you need to achieve the quality that you want. If the default doesn't suit, people can change it.

anyway, that's all irrelevant on the main question: how do we do this.

**Comment by robbietuk** [comment]
I think `number of subsets:=automatic` is the cleaner way to implement it into the parameter file. This also allows us to set it by default if `number of subsets:=` is not set. 

I think we need an intermediate variable for num_subsets (e.g. `num_subset_input_str`) and do something like:

In IterativeReconstruction.cxx->initialise_keymap()
``` 
this->parser.add_key("number of subsets", &this->num_subset_input_str);
```
In IterativeReconstruction.cxx->post_processing()
```
if ( num_subset_input_str.empty || num_subset_input_str == "automatic" )
  num_subset = compute_automatic_subset_num();
else
  num_subset = str_to_int(num_subset_input_str);
```

Obviously there are other methods like `ask_parameters()` that need to be addressed too but is a simple way to do it.
